File: app/src/main/java/com/shopcart/ui/fragments/HomeFragment.java
package com.shopcart.ui.fragments;


import android.app.Activity;
import android.os.Bundle;
import android.os.Handler;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.navigation.NavController;
import androidx.navigation.NavDestination;
import androidx.navigation.NavDirections;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewpager.widget.ViewPager;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.shopcart.R;
import com.shopcart.data.Repository;
import com.shopcart.data.models.User;
import com.shopcart.databinding.FragmentHomeBinding;
import com.shopcart.ui.adapters.CategoriesAdapter;
import com.shopcart.ui.adapters.HorizontalAdapter;
import com.shopcart.ui.adapters.SliderAdapter;
import com.shopcart.utilities.Tasks;

/**
 * A simple {@link Fragment} subclass.
 */
public class HomeFragment extends Fragment implements View.OnClickListener {
    FragmentHomeBinding binding;
    FirebaseFirestore firebaseFirestore;
    FirebaseStorage storage;
    StorageReference bannersRef;
    int sliderCurrentPage;
    Runnable sliderRunnable;
    Handler sliderHandler;
    int sliderDelay = 4000;
    private Activity activity;
    private FragmentManager fragmentManager;
    private FirebaseAuth mAuth;
    private User user;
    private SliderAdapter sliderAdapter;
    private HorizontalAdapter featuredAdapter;
    private HorizontalAdapter bestAdapter;
    private CategoriesAdapter categoriesAdapter;

    public HomeFragment() {
        // Required empty public constructor
    }


    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_home, container, false);

        if (isAdded()) {
            activity = getActivity();
        }

        Tasks.showSystemBars(activity);
        Tasks.defaultNavigationBar(activity);

        fragmentManager = getParentFragmentManager();
        user = Repository.getUserData();
        mAuth = FirebaseAuth.getInstance();
        firebaseFirestore = FirebaseFirestore.getInstance();
        storage = FirebaseStorage.getInstance();
        bannersRef = storage.getReference().child("banners");

        sliderHandler = new Handler();

        LinearLayoutManager categoriesLayoutManager = new LinearLayoutManager(activity, RecyclerView.HORIZONTAL, false);
        LinearLayoutManager featuredLayoutManager = new LinearLayoutManager(activity, RecyclerView.HORIZONTAL, false);
        LinearLayoutManager bestLayoutManager = new LinearLayoutManager(activity, RecyclerView.HORIZONTAL, false);


        binding.homeRecyclerFeatured.setLayoutManager(featuredLayoutManager);
        binding.homeRecyclerBest.setLayoutManager(bestLayoutManager);
        binding.homeRecyclerCategories.setLayoutManager(categoriesLayoutManager);

        binding.homeImgMenu.setOnClickListener(this);
        binding.homeImgCart.setOnClickListener(this);
        binding.homeTxtFeaturedMore.setOnClickListener(this);
        binding.homeTxtBestMore.setOnClickListener(this);

        sliderAdapter = new SliderAdapter(activity);
        featuredAdapter = new HorizontalAdapter(activity);
        bestAdapter = new HorizontalAdapter(activity);
        categoriesAdapter = new CategoriesAdapter(activity);

        binding.homePager.setAdapter(sliderAdapter);
        binding.homeTabDots.setupWithViewPager(binding.homePager, true);
        binding.homeRecyclerCategories.setAdapter(categoriesAdapter);
        binding.homeRecyclerFeatured.setAdapter(featuredAdapter);
        binding.homeRecyclerBest.setAdapter(bestAdapter);

        getData();

        binding.homePager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
            }

            @Override
            public void onPageSelected(int position) {
                sliderCurrentPage = position;
            }

            @Override
            public void onPageScrollStateChanged(int state) {
            }
        });

        sliderRunnable = new Runnable() {
            public void run() {
                if (sliderCurrentPage == sliderAdapter.getCount() - 1)
                    sliderCurrentPage = 0;
                else
                    sliderCurrentPage++;

                binding.homePager.setCurrentItem(sliderCurrentPage);
                sliderHandler.postDelayed(this, sliderDelay);
            }
        };


//        binding.button2.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View v) {
//                Tasks.deleteAccount(activity , mAuth , firebaseFirestore);
//            }
//        });


        return binding.getRoot();
    }

    private void getData() {
        sliderAdapter.setList(Repository.getBanners());
        categoriesAdapter.setList(Repository.getCategories());
        featuredAdapter.setList(Repository.getFeaturedProducts());
        bestAdapter.setList(Repository.getBestSellProducts());
    }

    @Override
    public void onClick(View v) {

        if (binding.homeImgMenu.equals(v)) {
            NavDirections action = HomeFragmentDirections.actionHomeFragmentToMenuFragment();
            NavController navController = NavHostFragment.findNavController(this);
            NavDestination navDestination = navController.getCurrentDestination();
            if (navDestination != null && navDestination.getId() == R.id.homeFragment)
                navController.navigate(action);

        } else if (binding.homeTxtFeaturedMore.equals(v)) {
            NavDirections action = HomeFragmentDirections.actionHomeFragmentToFeaturedFragment();
            NavController navController = NavHostFragment.findNavController(this);
            NavDestination navDestination = navController.getCurrentDestination();
            if (navDestination != null && navDestination.getId() == R.id.homeFragment)
                navController.navigate(action);

        } else if (binding.homeTxtBestMore.equals(v)) {
            NavDirections action = HomeFragmentDirections.actionHomeFragmentToBestSellFragment();
            NavController navController = NavHostFragment.findNavController(this);
            NavDestination navDestination = navController.getCurrentDestination();
            if (navDestination != null && navDestination.getId() == R.id.homeFragment)
                navController.navigate(action);

        } else if (binding.homeImgCart.equals(v)) {
            NavDirections action = HomeFragmentDirections.actionHomeFragmentToCartFragment();
            NavController navController = NavHostFragment.findNavController(this);
            NavDestination navDestination = navController.getCurrentDestination();
            if (navDestination != null && navDestination.getId() == R.id.homeFragment)
                navController.navigate(action);

        }
    }

    @Override
    public void onResume() {
        super.onResume();
        checkIfSignedIn();
        sliderHandler.postDelayed(sliderRunnable, sliderDelay);
    }

    private void checkIfSignedIn() {
        if (mAuth == null || mAuth.getCurrentUser() == null) {
            NavDirections action = HomeFragmentDirections.actionHomeFragmentToEntryFragments();
            NavController navController = NavHostFragment.findNavController(this);
            NavDestination navDestination = navController.getCurrentDestination();
            if (navDestination != null && navDestination.getId() == R.id.homeFragment)
                navController.navigate(action);
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        sliderHandler.removeCallbacks(sliderRunnable);
    }

}

File: app/src/main/java/com/shopcart/ui/fragments/HomeFragment.kt
package com.shopcart.ui.fragments

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.navigation.fragment.NavHostFragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.viewpager2.widget.ViewPager2
import com.google.android.material.tabs.TabLayoutMediator
import com.shopcart.R
import com.shopcart.databinding.FragmentHomeBinding
import com.shopcart.ui.adapters.CategoriesAdapter
import com.shopcart.ui.adapters.HorizontalAdapter
import com.shopcart.ui.adapters.SliderAdapter
import com.shopcart.ui.viewModels.MainViewModel
import com.shopcart.utilities.Resource
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

@AndroidEntryPoint
class HomeFragment : Fragment(), View.OnClickListener {
    private val TAG = HomeFragment::class.java.name

    private lateinit var binding: FragmentHomeBinding
    private val viewModel: MainViewModel by viewModels()

    private lateinit var sliderAdapter: SliderAdapter
    private lateinit var featuredAdapter: HorizontalAdapter
    private lateinit var bestAdapter: HorizontalAdapter
    private lateinit var categoriesAdapter: CategoriesAdapter

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        // Inflate the layout for this fragment
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_home, container, false)

        checkOnBoardingState()
        checkAuthenticationState()

        setupAdapters()

        binding.apply {
            homeRecyclerFeatured.layoutManager =
                LinearLayoutManager(requireContext(), RecyclerView.HORIZONTAL, false)
            homeRecyclerBest.layoutManager =
                LinearLayoutManager(requireContext(), RecyclerView.HORIZONTAL, false)
            homeRecyclerCategories.layoutManager =
                LinearLayoutManager(requireContext(), RecyclerView.HORIZONTAL, false)

            homePager.adapter = sliderAdapter
            homeRecyclerCategories.adapter = categoriesAdapter
            homeRecyclerFeatured.adapter = featuredAdapter
            homeRecyclerBest.adapter = bestAdapter

            TabLayoutMediator(homeTabDots, homePager) { tab, position -> }.attach()

            homePager.registerOnPageChangeCallback(object : ViewPager2.OnPageChangeCallback() {

                override fun onPageSelected(position: Int) {
                    super.onPageSelected(position)
                    // save position
                    viewModel.sliderCurrentPage = position
                }
            })

            homeImgMenu.setOnClickListener(this@HomeFragment)
            homeImgCart.setOnClickListener(this@HomeFragment)
            homeTxtFeaturedMore.setOnClickListener(this@HomeFragment)
            homeTxtBestMore.setOnClickListener(this@HomeFragment)
        }

        return binding.root
    }

    private fun setupSlider() {
        CoroutineScope(Dispatchers.Default).launch {
//            while (true) {
//                if (sliderAdapter.itemCount > 0) {
//                    // check for last item
//                    if (viewModel.sliderCurrentPage == sliderAdapter.itemCount - 1)
//                        viewModel.sliderCurrentPage = 0
//
//                    viewModel.sliderCurrentPage++
//                    binding.homePager.currentItem = viewModel.sliderCurrentPage
//
//                    Log.d(TAG, "slider in process")

//                //repeat Task Here
//                delay(2000)
//            }
        }
    }

    override fun onStart() {
        super.onStart()
        getData()
        setupSlider()
    }

    private fun setupAdapters() {
        sliderAdapter = SliderAdapter()
        featuredAdapter = HorizontalAdapter(HorizontalAdapter.OnClickListener { position, item ->
            // TODO item click listener
        })
        bestAdapter = HorizontalAdapter(HorizontalAdapter.OnClickListener { position, item ->
            // TODO item click listener
        })
        categoriesAdapter = CategoriesAdapter(CategoriesAdapter.OnClickListener { position, item ->
            // TODO item click listener
        })
    }

    private fun getData() {
        // Banners
        viewModel.banners.observe(this, Observer {
            when (it) {
                is Resource.Loading -> {
                    // TODO Do shimmer effect
                }

                is Resource.Success -> {
                    sliderAdapter.submitList(it.data)
                }

                is Resource.Error -> {

                }
            }
        })

        // Categories
        viewModel.categories.observe(this, Observer {
            when (it) {
                is Resource.Loading -> {
                    // TODO Do shimmer effect
                }

                is Resource.Success -> {
                    categoriesAdapter.submitList(it.data)
                }

                is Resource.Error -> {

                }
            }
        })

        // Featured Products
        viewModel.featuredProducts.observe(this, Observer {
            when (it) {
                is Resource.Loading -> {
                    Log.d(TAG, "Loading")
                    // TODO Do shimmer effect
                }

                is Resource.Success -> {
                    featuredAdapter.submitList(it.data)
                    for (product in it.data!!) {
                        Log.d(TAG, product.name)
                    }
                }

                is Resource.Error -> {
                    Log.d(TAG, it.message!!)
                }
            }
        })

        // Best Sell Products
        viewModel.bestSellProducts.observe(this, Observer {
            when (it) {
                is Resource.Loading -> {
                    // TODO Do shimmer effect
                }

                is Resource.Success -> {
                    bestAdapter.submitList(it.data)
                }

                is Resource.Error -> {

                }
            }
        })
    }

    override fun onClick(v: View) {
        when (v) {
            binding.homeImgMenu -> {
                val action = HomeFragmentDirections.actionHomeFragmentToMenuFragment()
                val navController = NavHostFragment.findNavController(this)
                val navDestination = navController.currentDestination
                if (navDestination != null && navDestination.id == R.id.homeFragment) navController.navigate(
                    action
                )
            }

            binding.homeTxtFeaturedMore -> {
                val action = HomeFragmentDirections.actionHomeFragmentToFeaturedFragment()
                val navController = NavHostFragment.findNavController(this)
                val navDestination = navController.currentDestination
                if (navDestination != null && navDestination.id == R.id.homeFragment) navController.navigate(
                    action
                )
            }

            binding.homeTxtBestMore -> {
                val action = HomeFragmentDirections.actionHomeFragmentToBestSellFragment()
                val navController = NavHostFragment.findNavController(this)
                val navDestination = navController.currentDestination
                if (navDestination != null && navDestination.id == R.id.homeFragment) navController.navigate(
                    action
                )
            }

            binding.homeImgCart -> {
                val action = HomeFragmentDirections.actionHomeFragmentToCartFragment()
                val navController = NavHostFragment.findNavController(this)
                val navDestination = navController.currentDestination
                if (navDestination != null && navDestination.id == R.id.homeFragment) navController.navigate(
                    action
                )
            }

        }
    }

    private fun checkOnBoardingState() {
        Log.d(TAG, "check onboarding ${viewModel.onBoardingState}")
        // Check if we need to display our OnBoarding Fragment
        if (!viewModel.onBoardingState) {
            // The user hasn't seen the OnBoardingFragment yet, so show it
            Log.d(TAG, "The user hasn't seen the OnBoardingFragment yet, so show it")

            openOnBoardingFragment()
        }
    }

    private fun checkAuthenticationState() {
        // Check if the user didn't sign in
        if (viewModel.firebaseAuth.currentUser == null ||
            viewModel.firebaseAuth.currentUser?.isEmailVerified == false
        ) {

            // The user hasn't signed in , so WelcomeFragment
            openWelcomeFragment()

        }
    }

    private fun openWelcomeFragment() {
        val action =
            HomeFragmentDirections.actionHomeFragmentToEntryFragments()
        val navController = NavHostFragment.findNavController(this)
        val navDestination = navController.currentDestination
        if (navDestination != null && navDestination.id == R.id.homeFragment) navController.navigate(
            action
        )
    }

    private fun openOnBoardingFragment() {
        Log.d(TAG, "opening onBoarding Fragment")
        val action =
            HomeFragmentDirections.actionHomeFragmentToOnBoardingFragment()
        val navController = NavHostFragment.findNavController(this)
        val navDestination = navController.currentDestination
        if (navDestination != null && navDestination.id == R.id.homeFragment) {
            Log.d(TAG, "open onBoarding")
            navController.navigate(
                action
            )
        }
    }
}